File: RBF/RBF.py
import numpy as np
from sklearn.cluster import KMeans
import pandas as pd

def trainRBF(xin, yin, p):

    # Função Radial Gaussiana
    def pdfnvar(x, m, K, n):
        if(n == 1):
            r = np.sqrt(int(K))
            px = (1/(np.sqrt(2*np.pi*r*r)))*np.exp(-0.5 * ((x-m)/(r))**2)
        else:
            px = ((1/np.sqrt((2*np.pi)**n * (np.linalg.det(K))))) * np.exp(-0.5*((x-m) @ (np.linalg.inv(K) @ np.transpose(x-m))))
        return px

    nSamples = xin.shape[0]     # Numero de amostras.
    nDimension = xin.shape[1]     # Dimensao de entrada.  

    xin = np.matrix(xin)  # garante que xin seja matriz
    yin = np.matrix(yin)  # garante que yin seja matriz

    #clusteriza os dados de entrada por meio do algorítmo K-médias
    xclust = KMeans(n_clusters=p).fit(xin)

    # Armazena vetores de centros das funções
    m = np.matrix(xclust.cluster_centers_)
    covlist = []
    
    # Estima matrizes de covarância para todos os centros
    for i in range(p):
        ici = np.where(xclust.labels_ == i)
        xci = xin[ici,]
        if nDimension == 1:
            covi = np.var(xci)
        else:
            covi = np.cov(xci[0], rowvar=False)
        covlist.append(covi)

    H =  np.zeros((nSamples, p))

    #calcula matriz H
    for j in range(nSamples):
        for i in range(p):
            mi = m[i,]
            covi = np.transpose(covlist[i]) + 0.001*np.identity(nDimension)
            aux = pdfnvar(xin[j,], mi, covi, nSamples)
            aux = np.asarray(aux)
            H[j][i] = aux[0]

    Haug = pd.DataFrame(H)
    Haug.insert(H.shape[1], H.shape[1], 1)
    Haug.to_numpy()

    W = (np.linalg.inv(Haug.T @ Haug) @ Haug.T) @ yin.T    # W<-( solve( t(Haug) %*% Haug) %*% t (Haug) ) %*% yin

    return [m, covlist, W, H]



# Calcula a saída da rede RBF
def YRBF(xin, modRBF):

    # Função Radial Gaussiana
    def pdfnvar(x, m, K, n):
        if(n == 1):
            r = np.sqrt(int(K))
            px = (1/(np.sqrt(2*np.pi*r*r)))*np.exp(-0.5 * ((x-m)/(r))**2)
        else:
            px = ((1/np.sqrt((2*np.pi)**n * (np.linalg.det(K))))) * np.exp(-0.5*((x-m) @ (np.linalg.inv(K) @ np.transpose(x-m))))
        return px

    nSamples = xin.shape[0]     # Numero de amostras.
    nDimension = xin.shape[1]     # Dimensao de entrada.  

    m = modRBF[0]
    covlist = modRBF[1]
    p = len(covlist) # Numero de funções radiais
    W = modRBF[2]

    H = np.zeros((nSamples, p))

    #calcula matriz H
    for j in range(nSamples):
        for i in range(p):
            mi = m[i,]
            covi = np.transpose(covlist[i]) + 0.001*np.identity(nDimension)
            aux = pdfnvar(xin[j,], mi, covi, nSamples)
            aux = np.asarray(aux)
            H[j][i] = aux[0]
    
    Haug = pd.DataFrame(H)
    Haug.insert(H.shape[1], H.shape[1], 1)
    Haug.to_numpy()

    Yhat = Haug @ W

    return Yhat

File: RBF/RBF.R
treinaRBF<-function(xin, yin, p) {
####### Fun ̧ca ̃o radial Gaussiana 
	pdfnvar<-function(x,m,K,n)
	{
		if(n==1) # nolint
		{
			r<-sqrt(as.numeric(K))
			px<-(1/(sqrt(2*pi*r*r)))*exp(-0.5 * ((x-m)/(r))^2)
		}
		else px<-((1/(sqrt((2*pi)^n*(det(K)))))*exp(-0.5*(t(x-m) %*% (solve(K)) %*% (x-m))))
		return (px)
	}
	##########################
	N<-dim(xin)[1] # n u ́ m e r o d e a m o s t r a s
	n<-dim(xin)[2] # dimens ̃ao de entrada (deve ser maior que 1)
	xin<-as.matrix(xin) # garante que xin seja matriz
	yin<-as.matrix(yin) # garante que yin seja matriz
	xclust<-kmeans(xin ,p)

	# Armazena vetores de centros das fun ̧c ̃oes .
	m<-as.matrix(xclust$centers) # nolintpr
	covlist<-list()

	# Estima matrizes de covariância para todos os centros .
	for ( i in 1:p)
	{
		ici<-which(xclust$cluster == i)
		xci<-xin[ici,] 
		if (n ==1)
			covi<-var(xci)
		else covi<-cov(xci)
		covlist[[i]]<-covi
	}

	H<-matrix(nrow = N,ncol = p)	

	# Calcula matriz H
	for (j in 1:N)
	{
		for ( i in 1:p)
		{
			mi<-m[i,]
			covi<-covlist[i] 
			covi<-matrix(unlist(covlist[i]),ncol=n,byrow =T) + 0.001*diag(n)
			H[j,i]<-pdfnvar(xin[j,] ,mi, covi, n)
		} 
	}
	Haug<-cbind (1 ,H)
	W<-( solve( t(Haug) %*% Haug) %*% t (Haug) ) %*% yin
	return(list(m,covlist,W,H))
}



YRBF<-function(xin, modRBF)
{
	####### Fun ̧c ̃ao radial Gaussiana
	pdfnvar<-function(x,m,K,n)
	{
		if (n==1) {
			r<-sqrt(as.numeric(K))
			px<-(1/(sqrt(2*pi*r*r)))*exp(-0.5*((x-m)/(r))^2)
		}
		else px<-((1/(sqrt((2*pi)^n*(det(K)))))*exp(-0.5*(t(x-m) %*% (solve(K)) %*% (x-m))))
		return (px)
	}

	##########################
	N<-dim(xin)[1] # numero de amostras
	n<-dim(xin)[2] # dimens ̃ao de entrada (deve ser maior que 1)
	m<-as.matrix(modRBF[[1]])
	covlist<-modRBF[[2]]
	p<-length(covlist) # Nu ́mero de fun ̧co ̃es radiais 
	W<-modRBF[[3]]

	xin<-as.matrix(xin) # garante que xin seja matriz
	# yin<-as.matrix(yin) # garante que yin seja matriz
	
	H<-matrix(nrow = N,ncol = p)
	# Calcula matriz H
	for (j in 1:N) {
		for ( i in 1:p) {
			mi<-m[i,]
			covi<-covlist[i]
			covi<-matrix(unlist(covlist[i]),ncol=n,byrow =T) +0.001*diag(n)
			H[j,i]<-pdfnvar(xin[j,], mi, covi, n) 
		}
	}
	Haug<-cbind(1,H)
	Yhat<-Haug %*% W

	return (Yhat)
}